### PR TITLE
feat(connectors): G3.7-T8 Hetzner Robot spec ingestion + operator-review + 10-op read core

### LINKED ISSUE #849: Task: G3.7-T8 Hetzner Robot spec ingestion + operator-review + enable ~10-op read-only core (Robot OpenAPI via G0.7)
Parent initiative: #370
Parent goal: #214
Depends-on: #846 (HetznerRobotConnector skeleton must be registered before ingested rows dispatch)

## Target

- Ingest: `docs:hetzner-robot-2026-04/<robot-api>.yaml` (Hetzner Robot OpenAPI) via the G0.7 pipeline.
- Add: `backend/src/meho_backplane/connectors/hetzner_robot/core_ops.py` (read-only core curation; mirror [`connectors/harbor/core_ops.py`](../blob/main/backend/src/meho_backplane/connectors/harbor/core_ops.py))
- Add: `backend/tests/test_connectors_hetzner_robot_core_ops.py`
- Add: `docs/cross-repo/g37-hetzner-canary.md` (ingestion canary; mirror [`g35-nsx-canary.md`](../blob/main/docs/cross-repo/g35-nsx-canary.md))
- Touches: `endpoint_descriptor` + `operation_group` rows under `connector_id="hetzner-rest-2026-04"`; operator review queue.
- Pipeline: `register_ingested_operations()` at [`operations/ingest/register_ingested.py:247`](../blob/main/backend/src/meho_backplane/operations/ingest/register_ingested.py#L247); runbook [`docs/cross-repo/connector-ingestion.md`](../blob/main/docs/cross-repo/connector-ingestion.md).

## Context

Per [CLAUDE.md](../blob/main/CLAUDE.md) postulates 1–2, Hetzner Robot is **generic-ingested** — ops enter `endpoint_descriptor` via G0.7 OpenAPI ingestion. G0.7 parses the spec, embeds, proposes LLM groups with `when_to_use`, and stages for operator review. The consumer's `docs/hetzner-robot-2026-04/` shelf already carries the spec.

## Current state

`HetznerRobotConnector` resolves (#846) but no `endpoint_descriptor` rows exist for `hetzner-rest-2026-04`.

## Desired state

The Robot spec is ingested. A **read-only v0.5 core subset (~10 ops, per #370 DoD)** is operator-reviewed and enabled — illustrative, finalized at review: `hetzner-robot.about` · `.server.list` · `.server.info` · `.ip.list` · `.subnet.list` · `.vswitch.list` · `.vswitch.info` · `.failover.list` · `.rdns.list` · `.ssh_key.list`. Each enabled group carries a reviewed `when_to_use`; each op reviewed `llm_instructions`. `server.list` returns a JSONFlux handle.

## Acceptance criteria

- [ ] The Robot spec ingests cleanly (zero unresolved `$ref`s; body-hash-skip idempotent) — reproducible via `meho connector ingest --spec docs:hetzner-robot-2026-04/<robot-api>.yaml`.
- [ ] ~10 read-only core ops enabled under `connector_id="hetzner-rest-2026-04"`; everything else `staged`. Verifiable via `search_operations`.
- [ ] Every enabled op has reviewed `llm_instructions`; every group a reviewed `when_to_use` (assert non-placeholder).
- [ ] `server.list` returns a JSONFlux handle; `result_describe`/`result_query` resolve.
- [ ] `docs/cross-repo/g37-hetzner-canary.md` records the run.
- [ ] `ruff` + `mypy` clean; `pytest -n auto backend/tests/test_connectors_hetzner_robot_core_ops.py` passes.

## Out of scope

- The connector class / auth / no-retry-on-401 / form helper — #846.
- Hetzner Robot writes — `staged`, never enabled in v0.2.
- CLI verbs / E2E / onboarding doc — T9.

## References

- Precedent: G3.5 NSX ingestion [#614](https://github.com/evoila/meho/issues/614), [`connectors/harbor/core_ops.py`](../blob/main/backend/src/meho_backplane/connectors/harbor/core_ops.py), [`g35-nsx-canary.md`](../blob/main/docs/cross-repo/g35-nsx-canary.md); pipeline [`operations/ingest/register_ingested.py`](../blob/main/backend/src/meho_backplane/operations/ingest/register_ingested.py), [`docs/cross-repo/connector-ingestion.md`](../blob/main/docs/cross-repo/connector-ingestion.md).
- Hetzner Robot Webservice: https://robot.hetzner.com/doc/webservice/en.html
- Best-practices anchors (by pack name): **python_best_practices** — Pydantic v2 descriptor schemas; **ai_engineering_best_practices** — `when_to_use`/`llm_instructions` discipline.

### PR DESCRIPTION
**Pull request by kr3s0**
## Summary

- Add `backend/src/meho_backplane/connectors/hetzner_robot/core_ops.py` — the curated 10-op read-only v0.2 core for the Hetzner Robot Webservice connector: `ROBOT_CORE_GROUPS` (4 groups with operator-reviewed `when_to_use`), `ROBOT_CORE_OPS` (10 GET ops with `llm_instructions`), `apply_robot_core_curation()`, and `classify_robot_op()`. Mirrors `connectors/harbor/core_ops.py`.
- Update `connectors/hetzner_robot/__init__.py` to export the new core_ops symbols so operators can `from meho_backplane.connectors.hetzner_robot import apply_robot_core_curation`.
- Add `backend/tests/test_connectors_hetzner_robot_core_ops.py` — 34 tests covering the path classifier, llm_instructions/when_to_use completeness, and the `apply_robot_core_curation` substrate integration (SQLite, 10-op enable invariant, non-core-op disable, audit rows).
- Add `docs/cross-repo/g37-hetzner-canary.md` — operator runbook for the G0.7 ingest + curation + verify cycle. Update `docs/codebase/connectors-hetzner-robot.md` to reflect T8 additions.

Key note: `ROBOT_PRODUCT="hetzner"` (not `"hetzner-robot"`) — `parse_connector_id("hetzner-rest-2026.04")` extracts the first hyphen-segment of `impl_id="hetzner-rest"`, yielding `"hetzner"`. Same discrepancy as SDDC Manager; documented with a cross-reference.

## Test plan

- [x] `ruff check` — clean
- [x] `ruff format --check` — clean
- [x] `mypy src/meho_backplane/connectors/hetzner_robot/ --ignore-missing-imports` — clean
- [x] `pytest tests/test_connectors_hetzner_robot_core_ops.py -v` — 34 passed
- [x] Classifier: every curated `GET:/path` op_id resolves to its declared `group_key`; non-GET and uncurated paths return `"none"`
- [x] `apply_robot_core_curation` enables exactly 10 ops; non-core ops in curated groups stay `is_enabled=False`; all 4 groups land `review_status='enabled'`; audit rows written
- [x] `llm_instructions` on all 10 ops: `when_to_call` / `output_shape` / `next_step` all non-empty strings
- [x] `GET:/server` (server.list) op carries JSONFlux handle note in `llm_instructions.next_step`

## Out of scope

- T9: CLI verbs + E2E + onboarding doc
- Hetzner Robot write ops — stay `staged`, never enabled in v0.2
- Env-gated full spec ingest canary (requires spec file on CI runner — follow-up)

Closes #849

## Release Notes

* **New Features**
  * Added curated read-only core operations support for Hetzner Robot connector, enabling a pre-defined subset of 10 operations across 4 operational groups with operator-configurable instructions.

* **Documentation**
  * Added connector architecture documentation for Hetzner Robot.
  * Added operator runbook for Hetzner Robot canary deployment and curation workflow.

* **Tests**
  * Added comprehensive test coverage for core operations classification, curation logic, and integration effects.

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/919?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by kr3s0** [comment]
@coderabbitai review

File: backend/tests/test_connectors_hetzner_robot_core_ops.py [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+"""
+
+from __future__ import annotations
+
+import uuid
+from collections.abc import Iterator
+from typing import Any
+
+import pytest
+from sqlalchemy import select
```
Comment (kr3s0) [inline backend/tests/test_connectors_hetzner_robot_core_ops.py:34]: **[🛑 B1]** — AC4 unmet: JSONFlux force-handle acceptance test missing

**What:** The task acceptance criterion "server.list returns a JSONFlux handle; result_describe/result_query resolve" (AC#4 in issue #849) is not met. No acceptance tests exist at `tests/acceptance/test_g37_robot_jsonflux_force_handle.py` or `tests/acceptance/_robot_canary_fixtures.py`. The canary doc (`g37-hetzner-canary.md`) references `~tests.acceptance._robot_canary_fixtures.ForceHandleReducer` which does not exist.

**Why it matters:** The harbor T8 precedent (#620) shipped `tests/acceptance/test_g35_harbor_jsonflux_force_handle.py` and `tests/acceptance/_harbor_canary_fixtures.py` for the identical acceptance criterion. The Robot T8 issue #849 has the same explicit AC. Merging without it leaves the JSONFlux dispatch-seam untested, which is the whole point of the criterion.

**Required change:** Add `backend/tests/acceptance/_robot_canary_fixtures.py` and `backend/tests/acceptance/test_g37_robot_jsonflux_force_handle.py` mirroring `_harbor_canary_fixtures.py` and `test_g35_harbor_jsonflux_force_handle.py`. The `ForceHandleReducer` pattern from harbor/nsx/sddc applies directly — seed a `HetznerRobotCanary` fixture with a mock `GET:/server` response (JSON array), install the `ForceHandleReducer`, dispatch `hetzner-robot.server.list`, and assert `result.handle` is a populated `ResultHandle`.

File: docs/cross-repo/g37-hetzner-canary.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+call `GET /server` (the cheapest authenticated endpoint) and propagate
+the auth_failed posture.
+
+## Operator procedure
+
+### Step 1 — ingest the spec
+
+```bash
+meho connector ingest \
+  --product hetzner-robot --version 2026.04 --impl hetzner-rest \
```
Comment (kr3s0) [inline docs/cross-repo/g37-hetzner-canary.md]: **[🛑 B2]** — `--product hetzner-robot` in CLI examples breaks the operator workflow

**What:** Both `meho connector ingest` examples in the operator procedure (Step 1) use `--product hetzner-robot`. `register_ingested_operations` stores rows with `product="hetzner-robot"`. When the operator then calls `apply_robot_core_curation`, it calls `ReviewService.get_review_payload("hetzner-rest-2026.04", tenant_id)` which calls `parse_connector_id("hetzner-rest-2026.04")` returning `product="hetzner"`. `load_groups` queries `WHERE product="hetzner"` — finds nothing — raises `ConnectorNotFoundError`.

**Why it matters:** An operator following this runbook will successfully ingest the spec (Step 1 completes) but will get a hard `ConnectorNotFoundError` at Step 3 (`apply_robot_core_curation`). The runbook is actively misleading. Compare: the NSX canary (`g35-nsx-canary.md`) uses `--product nsx` (matching `parse_connector_id("nsx-4.2")` = `"nsx"`); the VCFA canary uses `--product vcfa` (matching `parse_connector_id("vcfa-rest-9.0")` = `"vcfa"`).

**Required change:** In `docs/cross-repo/g37-hetzner-canary.md`, replace `--product hetzner-robot` with `--product hetzner` in both `meho connector ingest` code blocks (approximately lines 119 and 128).

```diff
-  --product hetzner-robot --version 2026.04 --impl hetzner-rest \
+  --product hetzner --version 2026.04 --impl hetzner-rest \
```

File: backend/src/meho_backplane/connectors/hetzner_robot/core_ops.py [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+    """
+    try:
+        method, path = op_id.split(":", 1)
+    except ValueError:
+        return "none"
+    if method != "GET":
+        return "none"
+    for prefix, group_key in ROBOT_PATH_RULES:
+        if path.startswith(prefix):
+            return group_key
```
Comment (kr3s0) [inline backend/src/meho_backplane/connectors/hetzner_robot/core_ops.py:253]: **[⚠️ M1]** — `classify_robot_op` uses raw `startswith`; hypothetical paths like `/ip_address` or `/serverless` would be misclassified (also flagged by CodeRabbit)

**What:** The `for prefix, group_key in ROBOT_PATH_RULES` loop at line 246 checks `path.startswith(prefix)`. A path like `/ip_address` would match the `/ip` prefix and be classified as `robot-networking` instead of returning `"none"`. The current Robot Webservice spec has no such collision, but future spec revisions could introduce one.

**Why it matters:** The harbor and NSX precedents use the same pattern, but Robot paths are shorter and more collision-prone (`/ip`, `/key`, `/server`) than Harbor paths (`/api/v2.0/…`). The fix is a one-liner and eliminates the class of future misclassifications entirely.

**Required change:**
```diff
-    for prefix, group_key in ROBOT_PATH_RULES:
-        if path.startswith(prefix):
-            return group_key
+    for prefix, group_key in ROBOT_PATH_RULES:
+        if path == prefix or path.startswith(f"{prefix}/"):
+            return group_key
```

File: docs/cross-repo/g37-hetzner-canary.md [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+   ([`meho_backplane.operations.ingest.parse_openapi`](../../backend/src/meho_backplane/operations/ingest/openapi.py))
+   ingests `hetzner-robot-2026-04/<robot-api>.yaml` (the Robot Webservice
+   OpenAPI spec, containing the full set of dedicated-server management
+   endpoints) under one `connector_id="hetzner-rest-2026.04"`.
+2. **Register.** T2's
+   [`register_ingested_operations`](../../backend/src/meho_backplane/operations/ingest/register_ingested.py)
+   bulk-upserts every parsed operation into the `endpoint_descriptor`
+   table under the
+   `(product="hetzner-robot", version="2026.04", impl_id="hetzner-rest")`
+   connector triple. The first ingest finds the hand-rolled
```
Comment (kr3s0) [inline docs/cross-repo/g37-hetzner-canary.md:39]: **[⚠️ M2]** — Step 2 description states `product="hetzner-robot"` where the actual row-level product key is `"hetzner"` (also flagged by CodeRabbit)

**What:** Line 38 describes the `endpoint_descriptor` connector triple as `(product="hetzner-robot", version="2026.04", impl_id="hetzner-rest")`. The actual stored product key is `"hetzner"` — the value `parse_connector_id("hetzner-rest-2026.04")` returns (`impl_id="hetzner-rest"`, `product=impl_id[:first_dash]="hetzner"`). This is the same discrepancy documented in `core_ops.py` for the SDDC Manager case.

**Why it matters:** An operator doing a manual DB diagnostic (`SELECT * FROM operation_group WHERE product=hetzner-robot`) will find no rows.

**Required change:**
```diff
-   `(product="hetzner-robot", version="2026.04", impl_id="hetzner-rest")`
+   `(product="hetzner", version="2026.04", impl_id="hetzner-rest")`
```

**Comment by kr3s0** [comment]
## /auto-review-pr summary (iteration 1)




**Anchor task:** #849 — G3.7-T8 Hetzner Robot spec ingestion + operator-review + enable ~10-op read-only core (Robot OpenAPI via G0.7)
**Surface area:** backend
**Best-practices packs audited:** python, ai_engineering
**Verdict:** REQUEST_CHANGES
**Head SHA reviewed:** `32eebf59899a9ad3bb48474b2fdb2bf414bee0d7`
**AI reviewer (CodeRabbit):** received in 0s (review was already posted at review start)

### Acceptance criteria

- [x] AC1: Robot spec ingests cleanly — env-gated (no spec file on CI); acknowledged out-of-scope per PR "Out of scope" note; canary doc provides manual procedure
- [x] AC2: ~10 read-only core ops enabled under `connector_id="hetzner-rest-2026-04"`; everything else `staged` — verified by `test_apply_robot_core_curation_enables_exactly_10_ops` and `test_apply_robot_core_curation_disables_non_core_ops_in_curated_groups`
- [x] AC3: Every enabled op has reviewed `llm_instructions`; every group a reviewed `when_to_use` — verified by `test_robot_core_ops_llm_instructions_all_populated` and `test_robot_core_groups_when_to_use_all_populated`
- ❌ AC4: `server.list` returns a JSONFlux handle; `result_describe`/`result_query` resolve — **not met**. No acceptance test. Harbor T8 (#620) shipped `tests/acceptance/test_g35_harbor_jsonflux_force_handle.py` + `tests/acceptance/_harbor_canary_fixtures.py` for the identical criterion. Robot T8 has neither. The canary doc at `docs/cross-repo/g37-hetzner-canary.md` references `tests.acceptance._robot_canary_fixtures.ForceHandleReducer` which does not exist.
- [x] AC5: `docs/cross-repo/g37-hetzner-canary.md` records the run — file added
- [x] AC6: `ruff` + `mypy` clean; `pytest -n auto backend/tests/test_connectors_hetzner_robot_core_ops.py` passes — all verified (34 tests pass)

### Project-board alignment

| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #849 | not on board | Closes | manual | AC4 missing | fix AC4, then close on merge |

Note: issue #849 is not on project board 16. PR body has `Closes #849`; the link is correct and the close will fire on merge once blockers are resolved.

### Predicted vs. actual diff

- ✅ `backend/src/meho_backplane/connectors/hetzner_robot/core_ops.py` — new file, mirrors harbor/nsx precedent
- ✅ `backend/src/meho_backplane/connectors/hetzner_robot/__init__.py` — exports the new symbols
- ✅ `backend/tests/test_connectors_hetzner_robot_core_ops.py` — 34 unit tests covering classifier, completeness, and substrate integration
- ✅ `docs/codebase/connectors-hetzner-robot.md` — updated with T8 additions
- ✅ `docs/cross-repo/g37-hetzner-canary.md` — operator runbook added
- 🕳️ `backend/tests/acceptance/test_g37_robot_jsonflux_force_handle.py` — **missing** (AC4 unmet)
- 🕳️ `backend/tests/acceptance/_robot_canary_fixtures.py` — **missing** (referenced in canary doc and core_ops.py docstring)
- 🕳️ `backend/tests/acceptance/test_g37_robot_dispatch_smoke.py` — **missing** (harbor T8 precedent shipped dispatch smoke + jsonflux force-handle)

### Integration verification

- [x] `ruff check` — clean
- [x] `ruff format --check` — clean
- [x] `mypy src/meho_backplane/connectors/hetzner_robot/ --ignore-missing-imports` — clean
- [x] `pytest backend/tests/test_connectors_hetzner_robot_core_ops.py` — 34 passed
- [⚠️ not run] migrations replay — no migrations in this PR
- [⚠️ not run] `docker compose up` — not required for this surface
- [⚠️ not run] JSONFlux dispatch smoke — acceptance tests missing (B1)

### Findings (inline — each is a merge blocker per team policy)

For the `/auto-implement-issue --continue-pr` punch list:

- 🛑 B1 — AC4 unmet: `tests/acceptance/test_g37_robot_jsonflux_force_handle.py` and `tests/acceptance/_robot_canary_fixtures.py` missing; required change: add Robot JSONFlux force-handle acceptance tests mirroring `test_g35_harbor_jsonflux_force_handle.py` + `_harbor_canary_fixtures.py`
- 🛑 B2 — Canary doc `--product hetzner-robot` in CLI commands (lines 119, 128) is wrong; would cause `apply_robot_core_curation` to fail with `ConnectorNotFoundError` because `ReviewService` calls `parse_connector_id("hetzner-rest-2026.04")` which returns `product="hetzner"`, but `register_ingested_operations` would have stored rows under `product="hetzner-robot"`; required change: replace `--product hetzner-robot` with `--product hetzner` in both `meho connector ingest` CLI example blocks in `g37-hetzner-canary.md`
- ⚠️ M1 — `classify_robot_op` uses raw `startswith` instead of segment-boundary matching; a hypothetical future path like `/ip_address` or `/serverless` would be misclassified; required change: `if path == prefix or path.startswith(f"{prefix}/"):` (also flagged by CodeRabbit)
- ⚠️ M2 — Canary doc line 38 text says `product="hetzner-robot"` where it should say `product="hetzner"` (the value `parse_connector_id` produces); required change: correct the inline description to `(product="hetzner", version="2026.04", impl_id="hetzner-rest")` (also flagged by CodeRabbit)

### Adjacent improvements (non-blocking, not posted inline)

- `test_g37_robot_dispatch_smoke.py` — harbor/nsx/sddc each have a dispatch smoke test (alongside the JSONFlux force-handle test). Once B1 is fixed and `_robot_canary_fixtures.py` exists, a dispatch smoke test would round out the acceptance suite. Not a T8 blocker but worth adding for consistency.

### AI reviewer (CodeRabbit) cross-check

- Adopted: 2 (CR-1 → M1 inline, CR-2 → M2 inline)
- Disputed: 0
- Out of scope: 0

### Docs

- [x] `docs/codebase/connectors-hetzner-robot.md` — updated (T8 additions documented)
- ❌ `docs/cross-repo/g37-hetzner-canary.md` — exists but contains two errors (B2 + M2 above)

### What worked well

- The 34-test suite is thorough and well-structured — it covers the 10-op count invariant, non-core-op disable, audit-row emission, `llm_instructions` completeness, `when_to_use` completeness, classifier boundary cases, and the exact-match assertion between enabled ops and `ROBOT_CORE_OPS`. The SQLite-via-autouse fixture mirrors the harbor precedent cleanly.
- The `ROBOT_PRODUCT="hetzner"` discrepancy vs. `HetznerRobotConnector.product="hetzner-robot"` is prominently documented in both the module docstring and `docs/codebase/connectors-hetzner-robot.md`, following the SDDC Manager precedent. This prevents future operators from being confused by the split.

**Review by kr3s0 — COMMENTED** [review]
## /auto-review-pr summary (iteration 2)




**Anchor task:** #849 — G3.7-T8 Hetzner Robot spec ingestion + operator-review + enable ~10-op read-only core (Robot OpenAPI via G0.7)
**Surface area:** backend
**Best-practices packs audited:** python, ai_engineering
**Verdict:** APPROVE
**Head SHA reviewed:** `e528a7a4f69770c852ffb95657407303dd3b636c`
**AI reviewer (CodeRabbit):** not received within window — CodeRabbit auto-reviews disabled on non-default base branch (`feat/846-hetzner-robot-connector`); only reviewed iter-1 commit `32eebf5`

### Acceptance criteria

- [x] AC1: Robot spec ingests cleanly — env-gated (no spec file on CI); acknowledged out-of-scope per PR; canary doc provides manual procedure
- [x] AC2: ~10 read-only core ops enabled under `connector_id="hetzner-rest-2026-04"`; everything else `staged` — verified by 34-test suite (`test_apply_robot_core_curation_enables_exactly_10_ops`, `test_apply_robot_core_curation_disables_non_core_ops_in_curated_groups`)
- [x] AC3: Every enabled op has reviewed `llm_instructions`; every group a reviewed `when_to_use` — verified by `test_robot_core_ops_llm_instructions_all_populated` and `test_robot_core_groups_when_to_use_all_populated`
- [x] AC4: `server.list` returns a JSONFlux handle — **now met**. `tests/acceptance/_robot_canary_fixtures.py` (460 lines) and `tests/acceptance/test_g37_robot_jsonflux_force_handle.py` (177 lines) added; acceptance test ran and passed (`1 passed, 1 warning in 8.95s`). The `ForceHandleReducer` dispatches `GET:/server`, asserts `handle.total_rows == 3` (matching `ROBOT_CANARY_SERVERS`), and validates `result_describe`/`result_query` seam readiness.
- [x] AC5: `docs/cross-repo/g37-hetzner-canary.md` records the run — file present
- [x] AC6: `ruff` + `mypy` clean; `pytest -n auto backend/tests/test_connectors_hetzner_robot_core_ops.py` passes — verified (34 passed + 1 acceptance passed)

### Project-board alignment

| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #849 | not on board | Closes | manual | ✅ all AC met | close on merge |

### Predicted vs. actual diff (iteration 2 delta)

- ✅ `backend/src/meho_backplane/connectors/hetzner_robot/core_ops.py` — M1 fixed: `path == prefix or path.startswith(f"{prefix}/")` at line 252
- ✅ `backend/tests/acceptance/_robot_canary_fixtures.py` — B1 fixed: 460-line fixture file added; mirrors `_harbor_canary_fixtures.py`
- ✅ `backend/tests/acceptance/test_g37_robot_jsonflux_force_handle.py` — B1 fixed: 177-line AC4 acceptance test added; mirrors `test_g35_harbor_jsonflux_force_handle.py`
- ✅ `backend/tests/acceptance/conftest.py` — 10-line update; exports `ROBOT_CANARY_OPERATOR_TENANT`, `IngestedRobotCanary`, `ingested_robot_canary`, `robot_acceptance_operator`
- ✅ `docs/cross-repo/g37-hetzner-canary.md` — B2 fixed: three corrections: line 38 (`product="hetzner"`), line 119 (`--product hetzner`), line 128 (`--product hetzner`)

### Iteration 1 findings — resolution status

| ID | Severity | Headline | Status |
|---|---|---|---|
| B1 | 🛑 Blocker | AC4 unmet: JSONFlux acceptance tests missing | ✅ Fixed — `_robot_canary_fixtures.py` + `test_g37_robot_jsonflux_force_handle.py` added; test passes |
| B2 | 🛑 Blocker | Canary doc `--product hetzner-robot` breaks operator workflow | ✅ Fixed — all 3 occurrences corrected to `--product hetzner` |
| M1 | ⚠️ Major | `classify_robot_op` raw `startswith` misclassifies `/ip_address` | ✅ Fixed — anchored to `path == prefix or path.startswith(f"{prefix}/")` |
| M2 | ⚠️ Major | Canary doc line 38 `product="hetzner-robot"` vs actual `"hetzner"` | ✅ Fixed as part of B2 correction |

### Integration verification

- [x] `ruff check` — clean (all 6 changed files)
- [x] `ruff format --check` — clean (6 files already formatted)
- [x] `mypy src/meho_backplane/connectors/hetzner_robot/ --ignore-missing-imports` — clean (4 source files)
- [x] `pytest tests/test_connectors_hetzner_robot_core_ops.py` — 34 passed in 4.01s
- [x] `pytest tests/acceptance/test_g37_robot_jsonflux_force_handle.py` — 1 passed in 8.95s
- [⚠️ not run] migrations replay — no migrations in this PR
- [⚠️ not run] `docker compose up` — not required for this surface

### Findings (inline — each is a merge blocker per team policy)

No findings. All iteration 1 blockers and majors are resolved.

### Adjacent improvements (non-blocking, not posted inline)

- The classifier test suite doesn't include boundary cases for the new anchored match (e.g., `GET:/ip_address → "none"`, `GET:/serverless → "none"`). The fix is correct and verified by code inspection; boundary-case parametrize entries would make the intent explicit. Recommend as a follow-up in T9 or a separate cleanup commit — not a merge blocker.
- `test_g37_robot_dispatch_smoke.py` (harbor/nsx/sddc each have a dispatch smoke test alongside the JSONFlux force-handle test) — noted in iter-1 adjacent improvements; still worth a follow-up once `_robot_canary_fixtures.py` fixture set is stabilized.

### AI reviewer (CodeRabbit) cross-check

CodeRabbit reviewed the iter-1 commit (`32eebf5`) only; auto-reviews are disabled for this PR's non-default base branch. No new review was posted on `e528a7a`. Iter-1 CodeRabbit findings (CR-1 → M1, CR-2 → M2) are both resolved.

- Adopted from iter-1 CR review: 2 (both resolved in iter-2 fix)
- New findings from bot: 0 (no new review)

### Docs

- [x] `docs/codebase/connectors-hetzner-robot.md` — verified unchanged from iter-1 (accurate for T8 additions)
- [x] `docs/cross-repo/g37-hetzner-canary.md` — corrected in iter-2 (B2 + M2 resolved)

### What worked well

- The fix commit is surgical and correct: 5 files, 665 lines added/changed, zero scope creep. All iter-1 findings addressed in a single commit with a descriptive message that cross-references each finding ID (B1, B2, M1).
- The `_robot_canary_fixtures.py` fixture file is production-quality: respx mocking for all 10 curated ops, RFC 6761 `.test.invalid` base URL, clean teardown via `reset_dispatcher_caches()`. The JSONFlux acceptance test exercises the real `call_operation` dispatch path (not a unit mock), which is exactly what AC4 requires.

**Review by damir-topic — APPROVED** [review]
## /auto-review-pr summary (iteration 3)




**Anchor task:** #849 — G3.7-T8 Hetzner Robot spec ingestion + operator-review + enable ~10-op read-only core
**Surface area:** backend
**Best-practices packs audited:** python, ai_engineering
**Verdict:** APPROVE
**Head SHA reviewed:** `96a6c2f`
**AI reviewer (CodeRabbit):** received (2 findings, both already resolved in current head — see cross-check); CodeRabbit check now reports Review completed (pass)

This iteration is the **first CI run** for the PR (triggered by merging current `main` into the branch via merge commit `96a6c2f`). Two scrutiny points were the focus: the auth-signature adaptation to #957's `Operator`-based `auth_headers`, and T7/T8 overlap with the skeleton already on `main` (#906). Both are clean.

### Acceptance criteria
- [x] **AC1** — Robot spec ingests cleanly: env-gated full-spec ingest is an explicit follow-up (no spec file on CI runner); the canary doc records the manual procedure. Out-of-scope per PR description and acknowledged in `g37-hetzner-canary.md`.
- [x] **AC2** — ~10 read-only core ops enabled, everything else staged: verified by `test_apply_robot_core_curation_enables_exactly_10_ops` + `test_apply_robot_core_curation_disables_non_core_ops_in_curated_groups`. The enable-group cascade correctly skips operator-overridden non-core ops via `operator_disabled_op_ids` → `excluded_op_ids` in `ReviewService.enable_group`.
- [x] **AC3** — Every enabled op has reviewed `llm_instructions`; every group a reviewed `when_to_use`: verified by `test_robot_core_ops_llm_instructions_all_populated` + `test_robot_core_groups_when_to_use_all_populated`.
- [x] **AC4** — `server.list` returns a JSONFlux handle: met via `tests/acceptance/test_g37_robot_jsonflux_force_handle.py` (real `call_operation` dispatch through the `ForceHandleReducer` seam, asserting `handle.total_rows == 3`). Passed under CI integration-testcontainers.
- [x] **AC5** — `docs/cross-repo/g37-hetzner-canary.md` records the run: present and accurate.
- [x] **AC6** — `ruff` + `mypy` clean; `pytest` passes: CI `Python (ruff + mypy + pytest)` green.

### Scrutiny point 1 — auth-signature adaptation (#957)
Confirmed clean. Every `auth_headers` override (including `hetzner_robot/connector.py:163`) uses the new `(self, target, operator: Operator)` signature. All `raw_jwt=` occurrences in this PR's new code are `Operator(raw_jwt=...)` **constructor** calls — `Operator` is a Pydantic `BaseModel` with a legitimate `raw_jwt: str = Field(repr=False)` field — not the removed `auth_headers(raw_jwt=...)` API. mypy-strict (CI) is the authoritative check and is green. Test fixtures pin `KEYCLOAK_*` + `VAULT_ADDR` so transitive `Settings` reads during `Operator` construction don't fail.

### Scrutiny point 2 — T7/T8 overlap (#906 skeleton)
Confirmed clean. The skeleton files (`connector.py`, `session.py`) are **not** in this PR's diff — they are pre-existing from T7 #906 (merged). This PR is purely additive: new `core_ops.py`, an additive `__init__.py` export block, new tests + canary fixtures + docs. No duplication, no clash; `__init__.py` and `acceptance/conftest.py` changes follow the established harbor/nsx/sddc pattern verbatim.

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #849 | Open | Closes | ✅ | ✅ all AC met | close on merge |

T9 (#934, CLI verbs + E2E + onboarding) is a correctly-separated open follow-up — no overlap with this PR's scope.

### Predicted vs. actual diff
- ✅ `connectors/hetzner_robot/core_ops.py` — curated 10-op core, classifier, `apply_robot_core_curation` (API-faithful to `ReviewService.edit_group`/`enable_group`/`edit_op`)
- ✅ `connectors/hetzner_robot/__init__.py` — additive export of core-ops symbols
- ✅ `tests/test_connectors_hetzner_robot_core_ops.py` — 34 behavioural tests against real `ReviewService` + SQLite
- ✅ `tests/acceptance/_robot_canary_fixtures.py` + `test_g37_robot_jsonflux_force_handle.py` + `conftest.py` — AC4 dispatch + JSONFlux seam
- ✅ `docs/codebase/connectors-hetzner-robot.md` + `docs/cross-repo/g37-hetzner-canary.md`

No missing, no undisclosed, no expected-but-wrong files.

### Integration verification (relied on CI per operational guardrails)
- [x] `Python (ruff + mypy + pytest)` — green (3m21s)
- [x] `Python (integration testcontainers)` — green (7m46s); exercises the JSONFlux force-handle acceptance test end-to-end
- [x] `CLI API snapshot freshness`, `Go`, `Helm`, `CodeQL`, `Semgrep SAST`, `TruffleHog`, `DCO`, `Analyze` — all green
- [x] `CodeRabbit` — Review completed
- ⚠️ not run locally: full test suite / Docker / testcontainers (per operational guardrails — relied on CI, which is authoritative)

### Findings (inline)
None. No code changes are required.

### CodeRabbit cross-check
CodeRabbit posted 2 findings, both against the **stale pre-merge commit `32eebf5`**, and both are **already resolved** in the current head `96a6c2f`:
- Adopted: 0
- Disputed / already-resolved: 2
  - `classify_robot_op` segment-boundary matching — current code already uses `path == prefix or path.startswith(f"{prefix}/")` (`core_ops.py:252`).
  - canary-doc row-level `product` value — current doc already reads `product="hetzner"` (`g37-hetzner-canary.md:38`, and CLI examples at lines 119/128 use `--product hetzner`).
- Out of scope: 0

### Adjacent improvements (non-blocking, not posted inline)
- The canary fixture monkeypatches the private `instance._credentials_loader` with `# type: ignore[attr-defined]` (`_robot_canary_fixtures.py:443`). Acceptable here — it mirrors the established `_harbor_canary_fixtures` pattern and is test-only. If a public credential-loader injection seam is ever added, the harbor/nsx/sddc/robot canary fixtures could migrate together.

### Docs
- [x] `docs/codebase/connectors-hetzner-robot.md` — updated to reflect T8 (core ops, curation fn, classifier); no .NET analogies.

### What worked well
- Behaviour-faithful tests: `apply_robot_core_curation` is exercised against a real `ReviewService` + SQLite (not mocked), seeding extra non-core ops to prove the "enable group but pin non-core disabled" needle, and asserting audit rows.
- The `ROBOT_PRODUCT="hetzner"` vs `HetznerRobotConnector.product="hetzner-robot"` discrepancy is thoroughly documented in code, fixtures, and docs, with the SDDC precedent cross-referenced — exactly the kind of footgun that needs the inline narrative it got.
